# 10 Feature Engineering 

## 4- Feature Extraction

### 1- Özellik Çıkrımı (Featur Extraction)

veri bilimi ve makine öğrenmesi alanlarından oldukça önemli olan feature engineering ile eş anlamlı olarak kullanılır. özellik çıkarımı ham veriden değişken türetmek, üretmektir. 1 yapısal verilerden değişkenler türetmek, 2- yapısal olmayan verilerden değşikenler türetmektir. yapısal verilerden değişken türetmek dediğimiz elimiz zaten belirli bir yapı olan mevcut değişkenler üzerinden veri, değişken türetmek diğeri ise yapısal olmayan yani görüntü gibi ses gibi yazı gibi verilerden veri türetmektir. iyi de neden böyle özellikler istiyoruz numeric veriler üzerinde çalışabiliyoruz excel tarzı verilerle çalışabiliyoruz ancak elimizde bir resim olduğundabir yazı ve ses olduğunda bazı öğrenme algoritmlarını kullanmak için bu yapısl olmayan şeyleri matematiksel şekilde anlaşılabilecek şeylere yani numeric temsillere dönüştürmemiz gerekmektedir. bundan dolayı yapısal olmayanları da dönüştürmek ve bunlardan özellik çıkarmak işlemiyle ilgileniyor oluruz.


* 1- yapısal olabilecek değişkenlerden özellikler çıkarma, değişken üretme durumu

örneğin elimde bir zaman değişkeni var bu değşikende yıl ay gün saat ve  dakika var bunlardan nasıl yeni değişkenler üretebiliriz. dolayısıyla tek olan bu değişkenden yıl, ay, gün, saat değşikeni çıkabilir. hatta bu günlerin haftanın hangi günlerine denk geldiğini ve gün isimlerini bulabiliriz. işte bu duruma yapısal olan bazı değişkenlerden yeni değişkenler türetmeye örnek olarak verilebilir.

ikinci örnek olarak elimde normalde isim değişkenleri olduğunda bu değişkenlere kardinal ölçülebilirliği olamayan şekilde yaklaşıyordum ancak isimlerin başındaki mrs, miss, mr, mlle gibi titlelar bana artık bir bilgi veriyor olabilir ben bunlara bakarak artık işlem yapabilirim. bu titlelar çünkü artık bilgi taşıyor olacaklar. gruplandırılabilir genellenebilir bir hal alacaktır durum.
bu titlelardan ne çıkabilir örn, cinsiyeti çıkabilir, yaşlarına ilişkin bilgi çıkabilir, eğer dr varsa kişilerin mesleği orrtaya çıkabilir. yani gözlemleneceği üzere anlamsız gibi duran bir değişkene dikkatli bakılarak anlamlı olabilecek yeni değişkenler türetilebilir. burada anlamlı derken ilgili değişkene göre ayırt edicilik oluşturup oluşturmamasıdır. yani burada aslında üretebileceğim tüm değişkenleri olası tüm senaryolarda üretmeliyim ancak bağımlı değişken durumuna da dikkat etmem gerekmektedir bunu da incelemem gerekmektedir.

örneğin elimizde bir metin var ve nlp yoluyla film öneri sistemi geliştirmek istiyorum. artık burada doağl dil işleme ve görüntülü veri işleme problemlerinin kapısındayız. bazı yapısal olan ve olmayan verilerden bahsettik şu anda elimde olan veri yapısal olmayan bir veri ben bunu bilgisayara yapısal olmadığı için anlatamıyorum. peki ne yapabilirim, örneğin bu film gibi yüzlerce filmin açıklamalarını alırız bu yüzlerce filmin açıklamlarında yer alan bütün eşsiz kelimeleri sütunlara koyarız örneğin bu filmde bir kelimeden kaç tane var? bu filmde diğer kelimeden kaç tane var diye bunları kategorileştirebiliriz. bakın burada ihtiyaç problem ve problemin çözümü, ben burada bir ölçüm yapmak istiyorum. farklı tekniklerle bu metin üzerinde çalışmak istiyorum. bu yüzden bunu matematiksel bir temsile dökmeliyim. bunu vektöre ile temsil edebilirim. hepsini temsil ettiğim için bir matris formatında bu temsil işlemi gerçekleşmiş. matematiksel temsil ihtiyacı üzerinde bazı yöntemler, algoritma ve benzeri yöntemler kullanıyor olacağım. dolayısıyla bu metni bu matris şeklinde sayılara dönüştürdüğümde üzerinde işlem yapabilir hale geldim. yani aslında burada yapmış olduğumuz özellik çıkarımıdır. burada özellikler nelerdir? her bir kelimelerdir. her bir kelimenin ilgili film açıklamasında kaç defa geçtiği bilgisi türetilmiştir.

(artık deep learningin giriş kapısındayız burada da)görsel bir veride de işlemler fotoğrafteki görseldeki renk yoğunluklarının renk değiişmlerinin farklarının yine dil modelinde olduğu gibi matematiksel olarak temsil edilmesi gerekmektedir. burada üretielcek olan featurelar da bunları üretmektedir. bu featureların da ayırt edicilikleri birçok veriyle beslendiğinde ortya çıkacaktır. işte bu featureları üretmek için özel yapılar kullanılır.

yani burada benim yapısal verilerde de yapısal olmayan verilerde de değişken türetebiliyor olmam lazım. ana amaç budur ve doğal dil işleme gibi görüntülü veri işleme gibi derin öğrenme yöntemlerinde de önemli kısım kırılım feature engineering kısmıdır.

### 2- Binary Featurs

bu konuda genel kapsayıcı kabul gören br literatür yoktur. çünkü özellik çıkarımı işi problemden probleme değişebilir. özelinde yöntemden yönteme de değişebilir. burada karşımıza gelebilecek isimlendirmeler neler olabilir: flag, bool, true-false

özetle burada yapacağımız işlemler 0 1 şeklinde var olan değişkenlerin üzerinden yeni değişkenler türetmektir. dikkat bu label veya bşnary encoding değildir. var olan değişkenin içinden böyle değişkenler türetmek yani yeni bir şey türetmekle ilgileniyoruz var olanları değiştirmekle değil. 

In [1]:
# veri setimi yüklüyorum. titanic

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
!pip install missingno
import missingno as msno
from datetime import date
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import LocalOutlierFactor #çok değişkenli ayıklama yöntemi bunkarı gptye sor
from sklearn.preprocessing import MinMaxScaler, LabelEncoder, StandardScaler, RobustScaler #standartlaştırma ve dönüştürme fonksiyonları




Defaulting to user installation because normal site-packages is not writeable


In [2]:
df = pd.read_csv("titanic.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

burada kabin değişkeni vardı bazı değerlerde nan bazı değerlerde nan olmama durumu vardı. şimdi öyle bir işlem yapmak istiyorum ki nan olan yerlere 1 nan olmayan yerlere 0 yazmak istiyorum. normalde kabin değişkeninde çok fazla eksik değer olduğu için silebilirdik. ama şunu merak ediyorum bu nan yerlere 0 diğerlerine 1 yazayım ve gözlemleyeyim bakayım mesela bağımlı değişken ile bakayım bakalım gerekten anlamsız mı sileyim mi diye fikir edinemye çalışıp düşünüyorum. işlemlere geçiyorum.

In [2]:
df["NWE_CABIN_BOOL"] = df["Cabin"].notnull().astype("int")
# cabin değişkeni notnull(null değil mi) yani dolu mu diye soruyorum sonuç olarak buradan true false değerleri geliyor olacak ve 
# bu true false değerleri 1 ve 0 ı temsil edeceğinden dolayı astype int diyerek bunları integrea çeviriyorum.

In [3]:
df.head() #new_cabin_bool isminde yeni bir değişkene atadığım için bu isimde bir değişken geldi
# nan olan yere 0 dedi dolu olan yere 1 dedi.
#şu anda anlamsız olabilecek bir değişken benim için artık daha anlamlı bir hale geldi.


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NWE_CABIN_BOOL  
0      0         A/5 21171   7.2500   NaN        S               0  
1      0          PC 17599  71.2833   C85        C               1  
2      0  STON/O2. 3101282   7.9250   NaN        S               0  
3      0            113803  53.1000  C123        S               1  
4      0            373450   8.0500   NaN        S               0

In [4]:
#şimdi de bağımlı değişkenime göre ortalama alacağım
#yeni oluşturdğum değişkene göre groupby alıp survivedın ortalamasını al diyorum
df.groupby("NWE_CABIN_BOOL").agg({"Survived": "mean"})

Survived
NWE_CABIN_BOOL          
0               0.299854
1               0.666667

In [5]:
#dikkat kabini dolu olanların hayatta kalma oranı %66 boş olanların ise %29
#yani nan olanlara göre daha yüksek.
#daha önceden benim için önemsiz gibi duran bu değişken artık benim için oldukça önemli bir hale geldi
# çünkü artık içerisinde önemki bir bilgi barındırıyor.
# artık şu yorumu yapabiliyorum artık bağımlı değişkenim survived açısından değerlendirdiğimde cabini dolu olan kişilerin hayatta kalma oranı boş olanlara göre 2 ka tdaha fazla
# bu durumda bunların arasında gerçekten bir ilişki olabilir mi diye düşünüyorum.
#buraya bi istatistiki test yapacağım. oran testi yapacağım.
# iki grubun oranını kıyaslama testi yapacağım.
#yeni oluşturduğum bir feature var bu featureın bağımlı değişken ile aralarındaki ilişkiyi merak ediyorum ve bunun için oran testi yapıyorum

In [6]:
from statsmodels.stats.proportion import proportions_ztest #ora ntstini import ettim

In [7]:
test_stat, pvalue = proportions_ztest(count = [df.loc[df["NWE_CABIN_BOOL"] == 1, "Survived"].sum(),
                                              df.loc[df["NWE_CABIN_BOOL"] == 0, "Survived"].sum()],
                                     
                                     nobs = [df.loc[df["NWE_CABIN_BOOL"] == 1, "Survived"].shape[0],
                                            df.loc[df["NWE_CABIN_BOOL"] == 0, "Survived"].shape[0]])

print("Test Stat = %.4f, p-value = %.4f" % (test_stat, pvalue))

# ilk kısımda kabin numarası 1 olanların survivedlarının sumını aldım yani kabin numarası olup hayatta olan kaç kişi var, 
#sonra kabin numarası 0 olanların da sumını aldım yani kabin numarası olmayıp hayatta kalan kaç kişi var. 
# oran testi yapmak için gerekli 2 parametre var birincisi başarı sayısı ikincisi ise gözlem sayısı
# dolayısıyla bu problem için başarının tamı kabin numarası olanlar için hayatta kalmaktır, kabin numarası olmayanalrın da hayatta kalan kişilerinin toplamını bulduk
# ve nobs yani gözlemlenme frekansları var. yani kabin numarası olanlar kaç kişi, kabin nnumarası olmayanlar kaç kişi
# bunları da nobs argümanına girdik.

Test Stat = 9.4597, p-value = 0.0000


dikkat oldukça güzel bir noktadayız. proportions z testinin sonucu şu şekildedir: p1 ve p2 oranları arasında fark yoktur der. p1 ve p2 oranları neyi ifade eder. iki grubun yani kabin numarası olan ve olmayanların hayatta kalıp kalmama durumunu ifade ediyor bu oranlar. ikisi arasında fark yoktur diyen h0 hipotezi p-value değeri 0.05 ten küçük olduğundan dolayı reddedilir. yani aralarında istatistiki olarak anlamlı bir farklılık var gibi gözüküyor.

bir değişken türettik ve bu değişken benim için anlamlı olabilir mi olmayabilir mi bilmiyordum. modelleme yapmadan bunu tam anlamıyla göremem ama bana bir fikir vermesi açısından bu oranlar frekanslar da bunun dağılımı da göz önünde bulundurulduğunda acaba birbirinden farklı mıdırı sordum ve farklıdır yanıtını aldım. iki kere dikkat burada çok değişkenli etkiyi bilmiyorum. burada iki tane değişkenin sanki ikisi birlikte oluşmuş gibi değerlendirerek getirdim. ama bu yapılar tek başına oluşmadı ki yani survived değişkeninin ortaya çıkışı burada ele aldığımız new kabin bool değişkenile ortaya çıkmadı ki am çıkmış da olabilir bilmiyorum. ve şu andaki kanaatim de çıkmış olabileceğine yönelik ama bilmiyorum. dolayısıyla neyi ifade etmeye çalışıyorum. çok değişkenli etkiyi bilmiyorum ben bu değişkenlerle modele soktuğumda bu çok değişkenli etkiyi gözlemleme imkanım olacak. o durumda bu değişkenin anlamlı olup olmadığını daha iyi değerlendiriyor olacağım(new kabin bool değişkenini). ama şu anda ilerlemek ya da bu değişkeni kabul edip etmemek için bir fikir üretmeye çalıştığımda yeterli ispatı yeterli delili buldum. ancak tekrar dikkat çok değişkenli etkiyi daha bilmediğim için bu benim için önemli bir değişkendir genellemesini henüz yapamıyorum, gözönünde bulunduruyorum sadece en azından kayda değer diyorum.

In [8]:
# şimdi binary feature oluşturmaya devam ediyorum

df.loc[((df["SibSp"] + df["Parch"]) > 0), "NEW_IS_ALONE"] = "NO" # bu veri setinde sibsp  ve parch olmak üzere iki değişken var bunlar yakın ve uzak olabilecek akrabalıkları ifade ediyor.yani kişinin gemide akrabalıklarını ifade eden iki değişken var.bunların toplamı diyorum eğer 0dan büyükse yeni bir değişken oluştur ve demek ki bu kişi yalnız değilmiş diyerek new_is_alone değişkenine no şeklinde yeni bir değer ata diyorum
df.loc[((df["SibSp"] + df["Parch"]) == 0), "NEW_IS_ALONE"] = "YES" # eğer bunların toplamı da 0a eşitse yes ata. burada olası featureları türetmeye çalışıyorum hala.
#yorumum şu belki kişi gemide tek olup olmamasına göre hayatta kalma refleksleri şekillenmişştir. gemide annesi babsı ya da çocuklarının olması onun hayatta kalma reflekslerini etkilemiş olabilir.
# yaptığım işlemde bir sonuç çıkmayabilir de ama denemem lazım. değişken üretmenin güç noktası da bu olası tüm değişkenleri üretmeye çalışıyprum.
#yaptığım işlemi survived açısından da değerlendirmem lazım

df.groupby("NEW_IS_ALONE").agg({"Survived": "mean"})

Survived
NEW_IS_ALONE          
NO            0.505650
YES           0.303538

evet aralarında bir fark var gibi gözüküyor. yalnız mı hayır yanında ailesi, akrabası var değişkenine baktığımızda hayatta kalma oranı daha fazla gibi duruyor. bu etken şans eseri ortaya çıkmış olabilir çok değişkenli başka etkiler de olabilir sadece ama ucundan fikir edinmek için baktığımda fark var gibi gözüküyor. çünkü yalnız olanların hayatta kalma oranı daha düşük. buraya da bir hipotez testi yapacağım.

In [9]:
test_stat, pvalue = proportions_ztest(count = [df.loc[df["NEW_IS_ALONE"] == "YES", "Survived"].sum(),
                                              df.loc[df["NEW_IS_ALONE"] == "NO", "Survived"].sum()],
                                     nobs = [df.loc[df["NEW_IS_ALONE"] == "YES", "Survived"].shape[0],
                                            df.loc[df["NEW_IS_ALONE"] == "NO", "Survived"].shape[0]])

print("Test Stat = %.4f, p-value = %.4f" % (test_stat, pvalue))

# yukarıda yaptığım işlemin aynısı 
# fark var gibi gözüküyor demiştik ve istatistiki olarak anlammlı bir fark mı bu diye bakıyorum

Test Stat = -6.0704, p-value = 0.0000


evet farklı gibi gözüküyor, h0 hipotezi reddedilir. h0 hipotezi der ki iki oran arasında fark yoktur ama reddildi varmış. bu neyi ifade eder, ilerlemek için iyi bir noktadayım, mutlak olarak hayatta kalmak için bu değişkenin etkili olduğunu göstermez ama bu değişkenin gözardı edilemeyeceğini gösterir, peki bunun etkisini daha etkili nasıl görürüz? modelleme bölümünde göreceğiz. bir model kurulduğunda oradaki etki daha net gözüküyor olacaktır.

### 3- Text Featurs

bu bölümde metinler üzerinden özellikler türetmeye çalışacağız.

In [2]:
df.head() #en sonunda gördüğüm bütün teknikleri veri önişleme feature engineering hepsini veri setine baştan uyguluyor olacağım.
#verimi ilk haliyel kullanmak istediğim için bu sefer kernel atmadım sadece veriyi okuttum

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
#buradaki metinlerden işlemler türetmek istiyorum. normalde kardinalitesi yüksek
#çöp olan bir veri olabilir name kısmı ancak bi kontrol ediyorum. çeşitli değişkenler türetmeye çalışacağım

In [4]:
# 1- letter count (harflari saydırma)
# örneğin buradaki harfleri saydırarak burada kaç tane harf olduğuna bakabiliriz
#burada yorum yapıyorum ve incelemeye çalışıyorum, neler çıkabileceğini görmeye çalışıyorum.
#örneğin bazıları kraliyet ailesi soyundandır dıdısının dıdısının dıdısının dıdısı olabilir.
#bu yüzden belki de kraliyet ailelerine yapılan bir ayrıcalık vardır ve bu buraya yansımış olabilir.
#normal hayatta o kadar ilginç değişkenler türüyor ki bu şekilde probleme yaratıcı olarak yaklaşmakta fayda vardır.


df["NEW_NAME_COUNT"] = df["Name"].str.len()  #string diyorum ve len metodunu uyguluyorum.

In [5]:
df.head() #bütün isimler sayıldı ve gözlemlendi ve değişken oluştu kaç harf varsa
#gözlemlenmiş oldu. harfleri saydım kelimeleri de sayabilirim

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_NAME_COUNT  
0      0         A/5 21171   7.2500   NaN        S              23  
1      0          PC 17599  71.2833   C85        C              51  
2      0  STON/O2. 3101282   7.9250   NaN        S              22  
3      0            113803  53.1000  C123        S              44  
4      0            373450   8.0500   NaN        S              24

In [6]:
# word count (kelime sayıları)
# apply ile lamba tanımlaması yapacağım

df["NEW_NAME_WORD_COUNT"] = df["Name"].apply(lambda x: len(str(x).split(" ")))
#bu lambda şunu yapsın ilgili ismi yakaladığında bu ismi komple stringe çevirsin daha sonra bunu split et boşluklarından ve kaç tane kelime varsa say

In [7]:
df.head() #kelimeleri saydırdık sonuçlar tabloya eklendi. meta bilgileri topladım artık daha detaya inebilirim

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_NAME_COUNT  \
0      0         A/5 21171   7.2500   NaN        S              23   
1      0          PC 17599  71.2833   C85        C              51   
2      0  STON/O2. 3101282   7.9250   NaN        S              22   
3      0            113803  53.1000  C123        S              44   
4      0            373450   8.0500   NaN        S              24   

   NEW_NAME_WORD_COUNT  
0                    4  
1                    7  
2                    3  
3                    7  
4                    4

In [8]:
# özel yapıları yakalamak

# örneğin isimlerin içerisindeki doktor ifadeleri vardı örneğin bu ifadeleri flagleyebilirim.
#aslında burada mesleklerle ilgili bir değişken yok ama bene tüm featureları türetmeliyim. ondan sonra modelleme basamağında değerlendirebilirim.
#bu arada featureları sadeleştirme basamağım da olmalı gözden kaçmamalı, mümkün olduğunca az sayıda değişken ile yüksek başarıya erişmeye çalışıyorum.
#o yüzden label encodingte elemeler yaptım. ama Feature engineering bölümünde öyle değişkenle türeyebilir ki bunlar bir şekilde hataları düşürür yani,
#başarılı tahminler elde etme imkanını sağlayabilir. bu sebeple mümkün olduğunca kurcalıyoruz.

df["NEW_NAME_DR"] = df["Name"].apply(lambda x: len([x for x in x.split() if x.startswith("Dr")]))
# name kolonunda gez ilgili satırı split et, split ettiğinde name satırlarında bulunan her bir değişkene liste halinde erişilebilir olacak her bir hücreye, sonra gezdiğin her bir kelime dr ile başlıyorsa bunu seç sonra lene al.(bu bir yorum belki bu işlem 10 farklı şekilde de yapılabilir.)


In [9]:
df.head() #headine baktığımda ilk etapta gözlemleyemedim. hızlı bir şekilde bakmak istiyorum

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_NAME_COUNT  \
0      0         A/5 21171   7.2500   NaN        S              23   
1      0          PC 17599  71.2833   C85        C              51   
2      0  STON/O2. 3101282   7.9250   NaN        S              22   
3      0            113803  53.1000  C123        S              44   
4      0            373450   8.0500   NaN        S              24   

   NEW_NAME_WORD_COUNT  NEW_NAME_DR  
0                    4            0  
1                    7            0  
2                    3            0  
3                    7            0  
4                    4            0

In [10]:
#doktora göre groupbya alıp survivedın ortalamasını alıyorum gözlemlemek ve incelemek için

df.groupby("NEW_NAME_DR").agg({"Survived": ["mean", "count"]})

Survived      
                mean count
NEW_NAME_DR               
0            0.38252   881
1            0.50000    10

In [11]:
# burada gözlemlendiği gibi doktor olanların hayatta kalma olasılığı daha yüksekmiş
# ancak bu kategorik değişken olduğu için frekanslarını gözlemlemeye çalışacağım, bu yüzden meanin yanına countu ekliyorum
# 10 tane dr varmış. 881 gözlemin olduğu yerde 10 tane dr olması ne kadar önemli yoruma açık olmakla birlikte en azından kayda değer bir sonuç elde ettik. 

### 4- Regex Featurs

bu bölümde regular exprection ile değişken türetme işlemlerini sürdüyor olacağız. isimlerin önündeki titleların önemli olduğunu biliyordum ve şimdi bunların üzerinde işlem yapıp bunları çekeceğim. 

In [12]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_NAME_COUNT  \
0      0         A/5 21171   7.2500   NaN        S              23   
1      0          PC 17599  71.2833   C85        C              51   
2      0  STON/O2. 3101282   7.9250   NaN        S              22   
3      0            113803  53.1000  C123        S              44   
4      0            373450   8.0500   NaN        S              24   

   NEW_NAME_WORD_COUNT  NEW_NAME_DR  
0                    4            0  
1                    7            0  
2                    3            0  
3                    7            0  
4                    4            0

In [13]:
#titleı incelediğimizde başında ve sonunda boşluk var ilk harfi büyük sonraki harfler
#küçük ve boşluktan önce nokta var. burada bir örüntü yakalamaya çalışıyorum filtrelemek için.
#bu yüzden buna regular exprection deniyor. örneğin Mr. Mrs. Miss. başında boşluk büyük ve küçük harfler ve sonrasında nokta.
#filtremi oluştururken büyük küçük harfleri seçmesini başında boşluk ve sonunda nokta olduğunu belirtmem gerekmektedir.

In [14]:
df["NEW_TITLE"] = df.Name.str.extract(" ([A-Za-z]+)\.", expand = False) #extract çıkar diyorum. peki nasıl çıkarmak istersin önünde boşluk olacak sonunda nokta olacak ve büyük ya da küçük harflerden oluşacak şekilde göreceğin bu ifadeleri yakala

In [15]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_NAME_COUNT  \
0      0         A/5 21171   7.2500   NaN        S              23   
1      0          PC 17599  71.2833   C85        C              51   
2      0  STON/O2. 3101282   7.9250   NaN        S              22   
3      0            113803  53.1000  C123        S              44   
4      0            373450   8.0500   NaN        S              24   

   NEW_NAME_WORD_COUNT  NEW_NAME_DR NEW_TITLE  
0                    4            0        Mr  
1                    7            0       Mrs  
2                    3            0      Miss  
3                    7            0       Mrs  
4                    4            0        Mr

In [16]:
df[["NEW_TITLE", "Survived", "Age"]].groupby(["NEW_TITLE"]).agg({"Survived": "mean", "Age": ["count","mean"]})
#tıtleı seç survivedı ve agei seç, tıtlea göre grupbya al sonra survivedın ortlamasını agein toplam sayısını ve ortalamsını al

Survived   Age           
               mean count       mean
NEW_TITLE                           
Capt       0.000000     1  70.000000
Col        0.500000     2  58.000000
Countess   1.000000     1  33.000000
Don        0.000000     1  40.000000
Dr         0.428571     6  42.000000
Jonkheer   0.000000     1  38.000000
Lady       1.000000     1  48.000000
Major      0.500000     2  48.500000
Master     0.575000    36   4.574167
Miss       0.697802   146  21.773973
Mlle       1.000000     2  24.000000
Mme        1.000000     1  24.000000
Mr         0.156673   398  32.368090
Mrs        0.792000   108  35.898148
Ms         1.000000     1  28.000000
Rev        0.000000     6  43.166667
Sir        1.000000     1  49.000000

öyle değerlerli bir bilgi elde ettik ki gözlenme ihtimali olmayan bir değişken yani bizim için dikkate alınma ihtimali olmayan bir değişkenden dahi çok değerli bir bilgi çıkardık. burada örnğin eskiden boş değerlere yaş değerlerinin ortalamasını atamıştık ama artık bunun böyle yapılmaması gerektiğini biliyorum. örneğin burada master sınıfının ortalama yaş değeri 4müş, miss sınıfında 146 kişi var ve yaş ortalaması 21miş, mr sınıfının yaş ortalaması 32ymiş. şu anda çok değerli bir bilgiye sahibim. veri setimde daha önce olmayan bir değişken türettim bu değişken üzerinden yaşarın ortalamasını aldığımda gördüm ki hani kategorik değişken kırılımında eksiklikleri atayalım demiştik ya buna hizmet edebilecek bir değişken türedi. şimdi bu kategorik değişken kırılımında ortalamaları doldursam daha mantıklı olacak. bu şekilde olası ihtimallerde değşiken türetmenin ne kadar faydalı olabileceği gözlemlenmiştir. yine burada agein frekansını aldık bu bize bir bilgi sunuyor ve survivedın ortalamasını aldık hayatta kalma kalmama durumunu ortaya koyması için bunu da gözlemlemek için yaptık.

### 5- Date Featurs

bu bölümde date değişkenleri, date featureları date özellikleri üretiyor olacağız

In [5]:
dff = pd.read_csv("course_reviews.csv")

dff.head()
#kursun puanlamaları var, puanlamaların tarihleri, kursa puan veren kiişnin kursa iye olma tarihi, ilerleme durumu, sorduğu sorular, yanıtlanan sorular gibi değişkenleri var
# amacım timesstamp değişkeninden yeni değişkenler türetmek.

Rating            Timestamp             Enrolled  Progress  \
0     5.0  2021-02-05 07:45:55  2021-01-25 15:12:08       5.0   
1     5.0  2021-02-04 21:05:32  2021-02-04 20:43:40       1.0   
2     4.5  2021-02-04 20:34:03  2019-07-04 23:23:27       1.0   
3     5.0  2021-02-04 16:56:28  2021-02-04 14:41:29      10.0   
4     4.0  2021-02-04 15:00:24  2020-10-13 03:10:07      10.0   

   Questions Asked  Questions Answered  
0              0.0                 0.0  
1              0.0                 0.0  
2              0.0                 0.0  
3              0.0                 0.0  
4              0.0                 0.0

In [6]:
dff.info() #ancak bir problemim var üzerinden yeni değişkenler üreteceğim değişkenin türü datetime olması gerekirken object gözüküyor.
# bu değişkenin tipini düzletmem gerekmektedir.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4323 entries, 0 to 4322
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rating              4323 non-null   float64
 1   Timestamp           4323 non-null   object 
 2   Enrolled            4323 non-null   object 
 3   Progress            4323 non-null   float64
 4   Questions Asked     4323 non-null   float64
 5   Questions Answered  4323 non-null   float64
dtypes: float64(4), object(2)
memory usage: 202.8+ KB


In [8]:
dff["Timestamp"] = pd.to_datetime(dff["Timestamp"], format = "%Y-%m-%d %H:%M:%S") #to_datetim metodunu kullanarak bunu gerçekleştirebiliriz,
#bana dönüştürmek istediğin object verini ver der ve bu değişkenin içinde çeşitli tarih zaman ifadeleri var ya bunların da ne sırada
#olduğunnu bana ver der, ofrmat kısmında da bu yazmaktadır year-month-day şeklinde 

In [9]:
dff.info() #aldığım hatada dattime modülüne dönüştürmek için veride saat için ayar yapmamı istedi hmsyi yazmadan hata verdi

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4323 entries, 0 to 4322
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Rating              4323 non-null   float64       
 1   Timestamp           4323 non-null   datetime64[ns]
 2   Enrolled            4323 non-null   object        
 3   Progress            4323 non-null   float64       
 4   Questions Asked     4323 non-null   float64       
 5   Questions Answered  4323 non-null   float64       
dtypes: datetime64[ns](1), float64(4), object(1)
memory usage: 202.8+ KB


In [10]:
#yıl değişkeni türetmek istiyorum

dff["year"] = dff["Timestamp"].dt.year # dt burada datetime modülünü kastetmektedir. dt.year diyerek yılı çek diyorum.

In [11]:
# ay değişkeni oluşturuyorum.
dff["year"] = dff["Timestamp"].dt.month

In [12]:
#bugünün tarihinden çıkarabilirim
#veri setindeki değişkenleri bugünün yılından çıkar diyebiliyorum ve yılların arkı geliyor
dff["year_diff"] = date.today().year - dff["Timestamp"].dt.year

In [15]:
#iki tarih arasındaki farkın ay cinsinden ifade edilmesi

dff["month_diff"] = (date.today().year - dff["Timestamp"].dt.year) * 12 + date.today().month - dff["Timestamp"].dt.month
#iki tarih arasındaki ay farkını hesaplamak istediğimizde öncelikle yıl farkını hesaplamalıyız sonra
#kaç yıl fark varsa onu 12 ile çarparak ay cinsine çevirmeliyiz. daha sonra iki trih arasındaki ay farkını almalıyız.
#date.today ile şu anın tarihi getiriliş sonra datadaki yıl alınmış ve bunların farkı alınmış ve 12 ile çarpılmış ve iki ay farkı ile toplanmış
# 2 * 12 = 24 ve geri kalan işlemler

In [16]:
dff.head()

Rating           Timestamp             Enrolled  Progress  Questions Asked  \
0     5.0 2021-02-05 07:45:55  2021-01-25 15:12:08       5.0              0.0   
1     5.0 2021-02-04 21:05:32  2021-02-04 20:43:40       1.0              0.0   
2     4.5 2021-02-04 20:34:03  2019-07-04 23:23:27       1.0              0.0   
3     5.0 2021-02-04 16:56:28  2021-02-04 14:41:29      10.0              0.0   
4     4.0 2021-02-04 15:00:24  2020-10-13 03:10:07      10.0              0.0   

   Questions Answered  year  year_diff  month_diff  
0                 0.0     2          2          34  
1                 0.0     2          2          34  
2                 0.0     2          2          34  
3                 0.0     2          2          34  
4                 0.0     2          2          34

In [17]:
#veri setindeki günlerin hangi gün olduğuna erişmek istersek

dff["day_name"] = dff["Timestamp"].dt.day_name()

In [19]:
dff.head() #yıl farkı ay farkı gün ismi ve (yılın gelmesi gerekiyordu.)

Rating           Timestamp             Enrolled  Progress  Questions Asked  \
0     5.0 2021-02-05 07:45:55  2021-01-25 15:12:08       5.0              0.0   
1     5.0 2021-02-04 21:05:32  2021-02-04 20:43:40       1.0              0.0   
2     4.5 2021-02-04 20:34:03  2019-07-04 23:23:27       1.0              0.0   
3     5.0 2021-02-04 16:56:28  2021-02-04 14:41:29      10.0              0.0   
4     4.0 2021-02-04 15:00:24  2020-10-13 03:10:07      10.0              0.0   

   Questions Answered  year  year_diff  month_diff  day_name  
0                 0.0     2          2          34    Friday  
1                 0.0     2          2          34  Thursday  
2                 0.0     2          2          34  Thursday  
3                 0.0     2          2          34  Thursday  
4                 0.0     2          2          34  Thursday

In [20]:
#date modülü içerisinde oldukça kapsamlı ihtiyaçlara göre metodlar bulunmaktadır. bakabilirsin.

### 6- Özellik Etkileşimleri (Feature Interaction)

özellik etkileşimi ne demektir? değişkenler birbiri ile etkileşime girmesi demektir. yani örneğin iki değişkenin çarpılmasıdır. iki değişkenin toplanmasıdır. bir değişkenin karesinin alınması küpünün alınması gibi değişkenlerle etkileşimde olmak demektir.

In [3]:
#örneğin yaş değişkeni ile pclassı çarpabiliriz.

df["NEW_AGE_PCLASS"] = df["Age"] * df["Pclass"] #bu interactionların teorisinde bir şey olmalı, bir şey ifade ediyor olmalı, örneğin bazı startupların tek görevi feature türetmektir. 
# burada olası tüm kombinasyonları denerler çarpıp köklerini alırlar bölerler toplarlar vs gibi ve bu yaptığımız işlem (önemli nokta burası) bir şey ifade ediyor olmalı

#burada yaş değişkeni ile pclassı çarpmak neyi ifade eder. age değişkeninin aynı standart formda olması varsayımı altında örneğin pclass 1 ile 3 arasında ya yaş değşikeni da 1 ile 3 arasında olsun
# ve burada yaşı küçük olup pclasıı 1 olan bir yolcu ile yaşı büyük olup yolculuk sınıfı 3 olan bir yolcunun hayatta kalma oranında fark olacaktır. hangi açıdan refah açısından, ister aşlesnden kaynakklı o gemiye binebiliyor olsun ister 
#kendi imkanlarıyla binebiliyor olsun dolayısıyla yaş değişkeni ile pclass değişkeninin çarpımı neticesinde gelen değerin küçük olması bir şey ifade ediyor olacaktır, büyüj olması da bir şey ifade ediyor olacaktır. mesela bu interaction, bu etkileşim teorik olarak bir şeye karşılık geliyor.
#belki de bu değişken modelleme esasında anlamlı bir yapı sahibi olacak. o yüzden üretmeye çalışıyorum. arkasındaki mantık yolcuların yaşlarının yolculuk sınıflarına göre refah durumlarıyla ilgili bir durum ortaya çıkarmak. işete genç yaşta 1. sınıfta yolculuk yapıyor ya da 3. sınıfta yolculuk yapıyor acaba refah durmumu düşük gibi bir bilgi taşıyor olabilir. 
# tekrar altınız çizmekte dayda var yaş değişkeninin bir standartlaştırılmaya tabi tutulması gerekebilir.

df["NEW_FAMILY_SIZE"] = df["SibSp"] + df["Parch"] + 1
# ya da akrabalık ilişkilerini toplayıp bir de kişiinin kendisini toplayınca family size adlı bir değişkene ulaşılabilir.


df.loc[(df["Sex"] == "male") & (df["Age"] <= 21), "NEW_SEX_CAT"] = "youngmale"  # cinsiyeti erkek olanları ve yaşı 21den küçük olanları yeni bir değişkene at ve bunları böyle isimlendir
#belirli kategorik değişken ya da sayısal değişkenlerin etkileşimleri yani etkileşim noktalrına da flagler atılabilir.burada değişkenler oluşturulabilir.

#aynı değişkene başka bir sınıf oluşturalım
df.loc[(df["Sex"] == "male") & ((df["Age"] > 21) & (df["Age"]) <= 50), "NEW_SEX_CAT"] = "maturemale" # new_sex_cat katgorisine  erkek olup yaşı 21den büyük ve 50den küçük olanları maturemale şeklinde isimlendir diyorum yani olgun erkekler olarak.

#daha yaşlı olanlara ise:
df.loc[(df["Sex"] == "male") & (df["Age"] > 50), "NEW_SEX_CAT"] = "seniormale" #şeklinde isimlendirebilirim.
# hala aynı değişkendeyim yakaladığım bazı yerlere göre isimlendirmeler yapıyorum.

#kadınlar için:
df.loc[(df["Sex"] == "female") & (df["Age"] <= 21), "NEW_SEX_CAT"] = "youngfemale"

#yukarıda yaptığım işlemlerin aynısını kadınlar için yapıyorum. olgun kadınlar:
df.loc[(df["Sex"] == "female") & ((df["Age"] > 21) & (df["Age"]) <= 50), "NEW_SEX_CAT"] = "maturefemale"

#daha olgun kadınlar

df.loc[(df["Sex"] == "female") & (df["Age"] > 50), "NEW_SEX_CAT"] = "seniorfemale"

In [4]:
df.head() #yeni kategorik değişkenim geldi ve ne ifade ediyor.

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  NEW_AGE_PCLASS  \
0      0         A/5 21171   7.2500   NaN        S            66.0   
1      0          PC 17599  71.2833   C85        C            38.0   
2      0  STON/O2. 3101282   7.9250   NaN        S            78.0   
3      0            113803  53.1000  C123        S            35.0   
4      0            373450   8.0500   NaN        S           105.0   

   NEW_FAMILY_SIZE   NEW_SEX_CAT  
0                2    maturemale  
1                2  maturefemale  
2                1  maturefemale  
3                2  maturefemale  
4                1    maturemale

In [5]:
df.groupby("NEW_SEX_CAT")["Survived"].mean() #surviveda göre groupby alıp ortlaamsını alıyorum.

#olgun kadınların hayatta kalma oranı ile olgun erkeklerin hayatta kalma oranı arasında oldukça fark gözlemleniyor.
#bu oldukça kaliteli bir değişken oldu ve kullanılabilir.
#maturefemalelerin hayatta kalma oranın daha yüksek olduğu gözlemleniyor.
# yeni kategorik değişkenler oluşturdum ve kadınlar ve erkekler arasında gözlemler yapabildim. yeni değişkenler türettim ve bunu etkileşimler üzerinden türetmiş oldum

NEW_SEX_CAT
maturefemale    0.730640
maturemale      0.194340
seniorfemale    0.941176
seniormale      0.127660
Name: Survived, dtype: float64